Debug build fails without -fpermissive with GCC 7.1 and Boost 1.65

Current master, Debug configuration fails with this error:
```
In file included from /home/users/khuck/src/phylanx/build-delphi-x86_64-Linux-gcc/boost-1.65.0/include/boost/icl/type_traits/identity_element.hpp:11:0,
                 from /home/users/khuck/src/phylanx/build-delphi-x86_64-Linux-gcc/boost-1.65.0/include/boost/icl/type_traits/unit_element.hpp:12,
                 from /home/users/khuck/src/phylanx/build-delphi-x86_64-Linux-gcc/boost-1.65.0/include/boost/icl/concept/interval.hpp:17,
                 from /home/users/khuck/src/phylanx/build-delphi-x86_64-Linux-gcc/boost-1.65.0/include/boost/icl/closed_interval.hpp:12,
                 from /home/users/khuck/src/phylanx/src/hpx/src/runtime/agas/addressing_service.cpp:48:
/home/users/khuck/src/phylanx/build-delphi-x86_64-Linux-gcc/boost-1.65.0/include/boost/icl/type_traits/type_to_string.hpp:56:12: error: partial specialization of ‘struct boost::icl::type_to_string<Unary<Type> >’ after instantiation of ‘struct boost::icl::type_to_string<std::__cxx11::basic_string<char> >’ [-fpermissive]
     struct type_to_string<Unary<Type> >
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/users/khuck/src/phylanx/build-delphi-x86_64-Linux-gcc/boost-1.65.0/include/boost/icl/type_traits/type_to_string.hpp:72:12: error: partial specialization of ‘struct boost::icl::type_to_string<Binary<Type1, Type2> >’ after instantiation of ‘struct boost::icl::type_to_string<std::__cxx11::basic_string<char> >’ [-fpermissive]
     struct type_to_string<Binary<Type1, Type2> >
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/src/runtime/agas/addressing_service.cpp
+++ b/src/runtime/agas/addressing_service.cpp
@@ -45,7 +45,6 @@
 #include <hpx/lcos/broadcast.hpp>
 
 #include <boost/format.hpp>
-#include <boost/icl/closed_interval.hpp>
 
 #include <cstddef>
 #include <cstdint>
@@ -60,69 +59,71 @@
 
 namespace hpx { namespace agas
 {
-struct addressing_service::gva_cache_key
-{ // {{{ gva_cache_key implementation
-  private:
-    typedef boost::icl::closed_interval<naming::gid_type, std::less>
-        key_type;
+    struct addressing_service::gva_cache_key
+    {    // {{{ gva_cache_key implementation
+    private:
+        typedef std::pair<naming::gid_type, naming::gid_type> key_type;
 
-    key_type key_;
+        key_type key_;
 
-  public:
-    gva_cache_key()
-      : key_()
-    {}
+    public:
+        gva_cache_key()
+          : key_()
+        {
+        }
 
-    explicit gva_cache_key(
-        naming::gid_type const& id_
-      , std::uint64_t count_ = 1
-        )
-      : key_(naming::detail::get_stripped_gid(id_)
-           , naming::detail::get_stripped_gid(id_) + (count_ - 1))
-    {
-        HPX_ASSERT(count_);
-    }
+        explicit gva_cache_key(
+                naming::gid_type const& id, std::uint64_t count = 1)
+          : key_(naming::detail::get_stripped_gid(id),
+                naming::detail::get_stripped_gid(id) + (count - 1))
+        {
+            HPX_ASSERT(count);
+        }
 
-    naming::gid_type get_gid() const
-    {
-        return boost::icl::lower(key_);
-    }
+        naming::gid_type get_gid() const
+        {
+            return key_.first;
+        }
 
-    std::uint64_t get_count() const
-    {
-        naming::gid_type const size = boost::icl::length(key_);
-        HPX_ASSERT(size.get_msb() == 0);
-        return size.get_lsb();
-    }
+        std::uint64_t get_count() const
+        {
+            naming::gid_type const size = key_.second - key_.first;
+            HPX_ASSERT(size.get_msb() == 0);
+            return size.get_lsb();
+        }
 
-    friend bool operator<(
-        gva_cache_key const& lhs
-      , gva_cache_key const& rhs
-        )
-    {
-        return boost::icl::exclusive_less(lhs.key_, rhs.key_);
-    }
+        friend bool operator<(
+            gva_cache_key const& lhs, gva_cache_key const& rhs)
+        {
+            return lhs.key_.second < rhs.key_.first;
+        }
 
-    friend bool operator==(
-        gva_cache_key const& lhs
-      , gva_cache_key const& rhs
-        )
-    {
-        // Direct hit
-        if(lhs.key_ == rhs.key_)
-            return true;
+        friend bool operator==(
+            gva_cache_key const& lhs, gva_cache_key const& rhs)
+        {
+            // Direct hit
+            if (lhs.key_ == rhs.key_)
+            {
+                return true;
+            }
 
-        // Is lhs in rhs?
-        if (1 == lhs.get_count() && 1 != rhs.get_count())
-            return boost::icl::contains(rhs.key_, lhs.key_);
+            // Is lhs in rhs?
+            if (1 == lhs.get_count() && 1 != rhs.get_count())
+            {
+                return rhs.key_.first <= lhs.key_.first &&
+                    lhs.key_.second <= rhs.key_.second;
+            }
 
-        // Is rhs in lhs?
-        else if (1 != lhs.get_count() && 1 == rhs.get_count())
-            return boost::icl::contains(lhs.key_, rhs.key_);
+            // Is rhs in lhs?
+            else if (1 != lhs.get_count() && 1 == rhs.get_count())
+            {
+                return lhs.key_.first <= rhs.key_.first &&
+                    rhs.key_.second <= lhs.key_.second;
+            }
 
-        return false;
-    }
-}; // }}}
+            return false;
+        }
+    }; // }}}
 
 addressing_service::addressing_service(
     parcelset::parcelhandler& ph

--- a/tests/performance/local/agas_cache_timings.cpp
+++ b/tests/performance/local/agas_cache_timings.cpp
@@ -18,7 +18,6 @@
 #include <hpx/util/histogram.hpp>
 
 #include <boost/program_options.hpp>
-#include <boost/icl/closed_interval.hpp>
 #include <boost/accumulators/accumulators.hpp>
 
 #include <algorithm>
@@ -37,52 +36,64 @@ typedef hpx::util::cache::entries::lfu_entry<hpx::agas::gva> gva_entry_type;
 struct gva_cache_key
 {
 private:
-    typedef boost::icl::closed_interval<hpx::naming::gid_type, std::less>
-        key_type;
+    typedef std::pair<hpx::naming::gid_type, hpx::naming::gid_type> key_type;
 
     key_type key_;
 
 public:
     gva_cache_key()
       : key_()
-    {}
+    {
+    }
 
-    explicit gva_cache_key(hpx::naming::gid_type const& id, std::uint64_t count)
+    explicit gva_cache_key(
+            hpx::naming::gid_type const& id, std::uint64_t count = 1)
       : key_(hpx::naming::detail::get_stripped_gid(id),
-             hpx::naming::detail::get_stripped_gid(id) + (count - 1))
+            hpx::naming::detail::get_stripped_gid(id) + (count - 1))
     {
         HPX_ASSERT(count);
     }
 
     hpx::naming::gid_type get_gid() const
     {
-        return boost::icl::lower(key_);
+        return key_.first;
     }
 
     std::uint64_t get_count() const
     {
-        hpx::naming::gid_type const size = boost::icl::length(key_);
+        hpx::naming::gid_type const size = key_.second - key_.first;
         HPX_ASSERT(size.get_msb() == 0);
         return size.get_lsb();
     }
 
     friend bool operator<(gva_cache_key const& lhs, gva_cache_key const& rhs)
     {
-        return boost::icl::exclusive_less(lhs.key_, rhs.key_);
+        return lhs.key_.second < rhs.key_.first;
     }
 
     friend bool operator==(gva_cache_key const& lhs, gva_cache_key const& rhs)
     {
+        // Direct hit
+        if (lhs.key_ == rhs.key_)
+        {
+            return true;
+        }
+
         // Is lhs in rhs?
         if (1 == lhs.get_count() && 1 != rhs.get_count())
-            return boost::icl::contains(rhs.key_, lhs.key_);
+        {
+            return rhs.key_.first <= lhs.key_.first &&
+                lhs.key_.second <= rhs.key_.second;
+        }
 
         // Is rhs in lhs?
         else if (1 != lhs.get_count() && 1 == rhs.get_count())
-            return boost::icl::contains(lhs.key_, rhs.key_);
+        {
+            return lhs.key_.first <= rhs.key_.first &&
+                rhs.key_.second <= lhs.key_.second;
+        }
 
-        // Direct hit
-        return lhs.key_ == rhs.key_;
+        return false;
     }
 };
 

